### ISSUE TITLE
[BUG] `compute_sliding_residuals` in `NaiveVariance` has overlap between train and test

### ISSUE THREAD

**Issue by bethrice44** [alan-turing-institute/sktime#2702]
**Describe the bug**

In [this line](https://github.com/alan-turing-institute/sktime/blob/df683766f4af7fd7fbe339fbf306d7a560827574/sktime/forecasting/naive.py#L635), a slice is taken to get train and test sets. The slicing behaviour appears to be inclusive for start and end for series so there is one element that is in both.

**To Reproduce**

```
from datetime import datetime
import pandas as pd

index = pd.date_range(datetime(2022, 5, 1, 9), datetime(2022, 5, 2, 8), freq="H")

y = pd.Series(index=index, data=[1]*len(index))

y_index = y.index

for id in y_index:
  y_train = y[:id]
  print(y_train)
  y_test = y[id:]
  print(y_test)
```

**Expected behavior**
Train and test sets should be exclusive. This can be achieved with:

```
y_train = y[:id][:-1]
```

**Versions**
<details>

<!--
Please run the following code snippet and paste the output here:
 
System:
    python: 3.8.12 (default, Oct 22 2021, 18:39:35)  [Clang 13.0.0 (clang-1300.0.29.3)]
executable: /Users/..../python
   machine: macOS-12.1-x86_64-i386-64bit

Python dependencies:
          pip: 22.0.3
   setuptools: 60.6.0
      sklearn: 1.0.2
       sktime: 0.11.3
  statsmodels: 0.13.2
        numpy: 1.21.5
        scipy: 1.7.3
       pandas: 1.4.1
   matplotlib: 3.5.1
       joblib: 1.1.0
        numba: 0.55.1
     pmdarima: None
      tsfresh: 0.19.0

**Comment by fkiraly** [comment]
What a mess!
Thanks for pointing this out, @bethrice44!

It would appear that the slicing behaviour is inconsistent when you compare datetime indexing to `iloc` indexing!
Compare this to `y[:5]` vs `y[5:]`, in which the split is non-overlapping.

I think there is a deeper problem here, with `pandas`, where this should be raised if it hasn't already been.

On the `sktime` side, we probably need to deal with this by case distinction in the slicing logic.

**Comment by fkiraly** [comment]
I confirm the issue on python 3.8 and windows.

**Comment by bethrice44** [comment]
> What a mess! Thanks for pointing this out, @bethrice44!
> 
> It would appear that the slicing behaviour is inconsistent when you compare datetime indexing to `iloc` indexing! Compare this to `y[:5]` vs `y[5:]`, in which the split is non-overlapping.
> 
> I think there is a deeper problem here, with `pandas`, where this should be raised if it hasn't already been.
> 
> On the `sktime` side, we probably need to deal with this by case distinction in the slicing logic.

On pandas, an [issue](https://github.com/pandas-dev/pandas/issues/36548) was raised, but it is apparently the documented behaviour, so no fix. 

> @IainStrachan this is the documented behaviour. Slicing with integers is positional indexing (see https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#slicing-ranges), while slicing with timestamps is using the labels. And then see here (https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#selection-by-label, the first paragraph after the three warning boxes) saying that label-based indexing includes both start and end point.

**Comment by bethrice44** [comment]
Could use something like this to check? 
```
        for id in y_index:
            forecaster = clone(forecaster)
            y_train = y[:id]  # subset on which we fit
            X_train = X[:id]
            y_test = y[id:]
            X_test = X[id:]
            if y[:id].max() == y[id:].min():
                y_test = y_test[1:]  # subset on which we predict
                X_test = X_test[1:]
```

**Comment by bethrice44** [comment]
Fixed in [this PR](https://github.com/alan-turing-institute/sktime/pull/2706) for `ConformalIntervals` using a new `get_slice` utility function which incorporates the conditional logic for timestamp/ period and integer indexes.

**Comment by fkiraly** [comment]
yes - as you proposed on slack, @bethrice44, `NaiveVariance` could (and probably should) be fixed the same way, i.e., using `get_slice`.

**Comment by bethrice44** [comment]
PR [here](https://github.com/alan-turing-institute/sktime/pull/2760) to fix this as described in comment.

**Comment by bethrice44** [comment]
Fixed and merged [here](https://github.com/alan-turing-institute/sktime/pull/2760)
